src/qibolab/instruments/emulator/pulse_simulator.py

```diff
@@ -0,0 +1,711 @@
+                                                        relaxation_time,
+                                                        fast_reset,
+                                                        acquisition_type,
+                                                        averaging_mode)
+
+        Returns:
+            dict: Qibolab results for AcquisitionType.DISCRIMINATION.
+
+        Raises:
+            TypeError: If execution_parameters.acquisition_type is AcquisitionType.RAW or AcquisitionType.INTEGRATION.
+        """
+        shape = [execution_parameters.nshots] + append_to_shape
+        results = {}
+        for ro_pulse in ro_pulse_list:
+            if execution_parameters.acquisition_type is AcquisitionType.DISCRIMINATION:
+                values = np.array(samples[ro_pulse.qubit]).reshape(shape)
+                processed_values = SampleResults(values)
+
+                if execution_parameters.averaging_mode is AveragingMode.SINGLESHOT:
+                    pass
+                elif execution_parameters.averaging_mode is AveragingMode.CYCLIC:
+                    processed_values = (
+                        processed_values.average
+                    )  # generaetes AveragedSampleResults
+
+            elif execution_parameters.acquisition_type is AcquisitionType.RAW:
+                raise TypeError("Emulator does not support raw measurement")
+            elif execution_parameters.acquisition_type is AcquisitionType.INTEGRATION:
+                raise TypeError("Emulator does not support integrated measurement")
+
+            results[ro_pulse.qubit] = results[ro_pulse.serial] = processed_values
+        return results
+
+    def run_pulse_simulation(
+        self,
+        sequence: PulseSequence,
+        instant_measurement: bool = True,
+    ) -> tuple[np.ndarray, list]:
+        """Simulates the input pulse sequence.
+
+        Args:
+            sequence (`qibolab.pulses.PulseSequence`): Pulse sequece to simulate.
+            instant_measurement (bool): Collapses readout pulses to duration of 1 if True. Defaults to True.
+
+        Returns:
+            tuple: Reduced density matrix of final state specified by readout qubits as well as the list of readout qubits in, both in little endian order.
+        """
+        # reduces measurement time to 1 dt to save simulation time
+        if instant_measurement:
+            for i in range(len(sequence)):
+                if type(sequence[i]) is ReadoutPulse:
+                    sequence[i].duration = 1
+
+        # extract waveforms from pulse sequence
+        channel_waveforms = ps_to_waveform_dict(
+            sequence,
+            self.platform2simulator_channels,
+            self.sampling_rate,
+            self.sim_sampling_boost,
+            self.runcard_duration_in_dt_units,
+        )
+
+        self.pulse_sequence_history.append(sequence.copy())
+        self.channel_waveforms_history.append(channel_waveforms)
```

Comment:
Avoid tracing the whole execution history inside the object.

`sequence` is user input, so if the user wants to keep the history, he could do it on its own.
Once you know the `pulse_sequence_history`, you should be able to compute the `channel_waveforms_history` just retrieving the settings from the `PulseSimulator` (or directly from the original configurations used to instantiate it).

Reply:
Fixed in d8b2ebd32e5d0bc392ee02eab54875ac842500ab